Add 3.12 overview docs. (#473)
faster-cpython/ideas: 3.12/interpreter_definition.md

# A higher level definition of the bytecode interpreter

## Abstract

The CPython interpreter is defined in C, meaning that the semantics of the
bytecode instructions, the dispatching mechanism, error handling, and
tracing and instrumentation are all intermixed.

This document proposes defining a custom C-like DSL for defining the 
instruction semantics and tools for generating the code deriving from
the instruction definitions.

These tools would be used to:
* Generate the main interpreter
* Generate the tier 2 interpreter
* Generate documentation for instructions
* Generate metadata about instructions, such as stack use.

Having a single definition file ensures that there is a single source
of truth for bytecode semantics.

Generating the interpreter decouples low-level details of dispatching
and error handling from the semantics of the instructions, resulting

> in more maintainable code and a potentially faster interpreter.

Reviewer: +1

Other tools that operate on bytecodes, like `frame.setlineno`
and the `dis` module, will be derived from the common semantic
definition, reducing errors.

## Motivation

> The bytecode interpreter of CPython has traditionally be defined as standard

Reviewer 2: ```suggestion
The bytecode interpreter of CPython has traditionally been defined as standard
```

C code, but with a lot of macros.
The presence of these macros and the nature of bytecode interpreters means
that the interpreter is effectively defined in a domain specific language (DSL).

Rather than using an ad-hoc DSL embedded in the C code for the interpreter,
a custom DSL should be defined and the semantics of the bytecode instructions,
and the instructions defined in that DSL.

This decouples low-level details of dispatching and error handling from the
semantics of the instructions, resulting in more maintainable code and a

> potentially faster interpreter.

Reviewer: This repeats some earlier text almost exactly.

It also provides the ability to create and check optimizers and optimization
passes from the semantic definition, reducing errors.

## Rationale

As we improve the performance of CPython, we need to optimize larger regions

> of code, use more complex optimizations and, ultimiteley, translate to machine

Reviewer 2: ```suggestion
of code, use more complex optimizations and, ultimately, translate to machine
```

code. 

> All of these step introduce the possibility of more bugs, and require more code

Reviewer 2: ```suggestion
All of these steps introduce the possibility of more bugs, and require more code
```

to be written. One way to mitigate this is through the use of code generators.
Code generators decouple the debugging of the code (the generator) from checking
the correctness (the DSL input).

For example, we are likely to want a new interpreter for the tier 2 optimizer
to be added in 3.12. That interpreter will have a different API, a different
set of instructions and potentially different dispatching mechanism.

> But the instructions it will interpreter will be built from the same building

Reviewer 2: ```suggestion
But the instructions it will interprete will be built from the same building
```

blocks as the instructions for the tier 1 (PEP 659) interpreter.

Rewriting all the instructions is tedious and error-prone, and changing the

> instructions is a maintainence haedache as both versions need to be kept in sync.

Reviewer 2: ```suggestion
instructions is a maintenance headache as both versions need to be kept in sync.
```

By using a code generator and using a common source for the instructions, or 

> parts of instructions, we can reduce the potential for errors considerably.

Reviewer: It also presents an opportunity for other projects, like Cinder, to generate custom interpreters with less effort.

## Specification

> This specification is at an early stage and is likely to change considerably.

> Syntax

Reviewer 3: You mention generating documentation as well. How does that fit into the syntax?

Also, I'm curious about how we specify the implicit pushes and pops that happen in e.g. exception handlers.

Author: Let's just wave our hands on the documentation part for now. We can generate the stack effect part of the `dis` docs.

------

Each op definition has a kind, a name, a stack and instruction stream effect,
and a piece of C code describing its semantics::
 
```
  definition:
    KIND NAME "(" (input ("," input))? "--" (output ("," output)*)? ")" "{" C-code "}"
    |
    KIND NAME "=" ( NAME | stream)+ ";"
    |
    KIND NAME "{" C-code "}"
 
  input:
    object | stream | array

  output:
    object | array

  object:
    NAME [":" NAME] [ "if" "(" C-expression ")" ]

  stream:
    (# | ## | ####) NAME

  array:
    object "[" NAME "]"

  family:
    FAMILY NAME = NAME ("," NAME)+ ";"

   file:
      (definition | family)+
```

The `KIND` must be one of:

> * `inst`: A normal instruction, as curently defined by `TARGET(NAME)` `in ceval.c`.

Reviewer 2: ```suggestion
* `inst`: A normal instruction, as currently defined by `TARGET(NAME)` in `ceval.c`.
```

* `op`: A part instruction from which other ops and instructions can be constructed.
* `super`: A super-instruction, such as `LOAD_FAST__LOAD_FAST`.

> `NAME` can be any ASCII identifier that is a C identifier and not a Python keyword.

Reviewer 2: ```suggestion
`NAME` can be any ASCII identifier that is a C identifier and not a C or Python keyword.
```
"struct" in the example below is not a Python keyword.

`foo_1` is legal. `$` is not legal, nor is `struct` or `class`.

The optional second name in an `object` is the type. It defaults to `PyObject *`.
The objects before the "--" are the objects on the stack at the start of the instruction.
Those after the "--" are the objects on the stack at the end of the instruction.

> The third form of `definition` is a transistion form to allow the original C code 

Reviewer 2: ```suggestion
The third form of `definition` is a transition form to allow the original C code 
```

definitions to be copied. It lacks information to generate anything other than the
interpreter, but is useful for initial porting of code.

The number of `#`s in a `stream` define how many codeunits are consumed from the
instruction stream.

The name `oparg` is pre-defined as a 32 bit value fetched from the instruction stream.

The C code may include special functions that are understood by the tools as
part of the DSL.

Those functions are:

* `DEOPT_IF(cond)`. Deoptimize if `cond` is met.
* `ERROR_IF(cond, handler)`. Jump to error handler if `cond` is true.
* `PEEK(n)`. The `n`<sup>th</sup> item on the stack, which is not pushed.

Variables can either be defined in the input, ouput, or in the C code.
Variables defined in the input may not be assigned in the C code.

> If an `ERROR_IF` occurs, all values will be removed from the stack.

Author: That's up to the definition, it is part of the semantics.
Which means that the code must be written so that the refcounts are correct at `ERROR_IF` and `DEOPT_IF` points.

        LOAD_INSTANCE_VALUE ####unused ;

    op LOAD_SLOT(owner, #index -- null if (oparg & 1), res) {
        char *addr = (char *)owner + index;
        res = *(PyObject **)addr;
        DEOPT_IF(res == NULL, LOAD_ATTR);
        Py_INCREF(res);
        null = NULL;
        Py_DECREF(owner);
    }

    inst LOAD_ATTR_SLOT = #counter CHECK_OBJECT_TYPE LOAD_SLOT ####unused;

    inst BUILD_TUPLE(items[oparg] -- tuple) {
        tuple = _PyTuple_FromArraySteal(items, oparg);
        ERROR_IF(tuple == NULL, error);
    }

    inst PRINT_EXPR {
        PyObject *value = POP();
        PyObject *hook = _PySys_GetAttr(tstate, &_Py_ID(displayhook));
        PyObject *res;
        if (hook == NULL) {
            _PyErr_SetString(tstate, PyExc_RuntimeError,
                                "lost sys.displayhook");
            Py_DECREF(value);
            goto error;
        }
        res = PyObject_CallOneArg(hook, value);
        Py_DECREF(value);
        ERROR_IF(res == NULL);
        Py_DECREF(res);
    }

    family load_attr = LOAD_ATTR, LOAD_ATTR_INSTANCE_VALUE, LOAD_SLOT ;
```

Generating the interpreter
==========================

> The generated C code for a single instruction includes a pre-amble and dispatch at the end

Reviewer 2: ```suggestion
The generated C code for a single instruction includes a preamble and dispatch at the end
```
